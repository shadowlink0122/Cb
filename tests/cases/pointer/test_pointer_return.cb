// 関数の戻り値としてのポインタテスト

// グローバル変数
int global_x = 10;
int global_y = 20;

// Test 1: グローバル変数へのポインタを返す
int* get_x_ptr() {
    return &global_x;
}

// Test 2: パラメータのポインタを返す（ポインタの転送）
int* forward_ptr(int* p) {
    return p;
}

// Test 3: 条件に応じて異なるポインタを返す
int* get_ptr(bool flag) {
    if (flag) {
        return &global_x;
    } else {
        return &global_y;
    }
}

// Test 4: nullptrを返す
int* get_null() {
    return nullptr;
}

int main() {
    // Test 1: グローバル変数へのポインタを返す
    println("Test 1: Return pointer to global");
    int* p1 = get_x_ptr();
    println(*p1);  // 10
    global_x = 30;
    println(*p1);  // 30
    
    // Test 2: ポインタの転送
    println("Test 2: Forward pointer");
    int local = 40;
    int* p2 = &local;
    int* p3 = forward_ptr(p2);
    println(*p3);  // 40
    *p3 = 50;
    println(local);  // 50
    
    // Test 3: 条件に応じたポインタ
    println("Test 3: Conditional pointer");
    int* p4 = get_ptr(true);
    println(*p4);  // 30 (global_x)
    int* p5 = get_ptr(false);
    println(*p5);  // 20 (global_y)
    
    // Test 4: nullptr
    println("Test 4: nullptr return");
    int* p_null = get_null();
    if (p_null == nullptr) {
        println("nullptr received");
    }
    
    return 0;
}
