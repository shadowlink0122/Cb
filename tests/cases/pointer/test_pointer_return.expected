Test 1: Return pointer to global
10
30
Test 2: Forward pointer
40
50
Test 3: Conditional pointer
30
20
Test 4: nullptr return
nullptr received
